File: leetcode/prefix_sum/238. Product of Array Except Self/py/main.py
class Solution:
    # Time O(N)
    # Space O(N)
    def productExceptSelf(self, nums):
        n = len(nums)
        left = [1] * n
        ans = [0] * n

        for i in range(1, n):
            left[i] = left[i - 1] * nums[i - 1]

        right = 1

        for i in range(n - 2, -2, -1):
            ans[i + 1] = left[i + 1] * right
            right = nums[i + 1] * right

        return ans

File: leetcode/prefix_sum/238. Product of Array Except Self/js/main.js
/*

  Алгоритм работы: 
  Суть решения проста, чтобы получить произведение в какой-либо точке 
  Необходимо знать префикс суммы левой половины от этого числа и в правой половины от этого числа
       
       nums           left             right 
  [1, 2, 3, 4]    [1, 1, 2, 6]    [24, 12, 4, 1]

                        [24, 12, 8, 6]

*/

// Time O(N)
// Space O(N)
const productExceptSelf = nums => {
  let n = nums.length;
  let dp = Array(n).fill(1);

  for (let i = 1; i < n; i++) {
    dp[i] = dp[i - 1] * nums[i - 1];
  }

  let right = 1;

  for (let i = n - 1; i >= 0; i--) {
    dp[i] = dp[i] * right;
    right = nums[i] * right;
  }

  return dp;
};
